python/cudf/bindings/groupby/groupby.pyx

```diff
@@ -0,0 +1,355 @@
+    df = cudf.DataFrame(dict(zip(range(len(cols)), cols)))
+    if columns is not None:
+        df.columns = columns
+    return df
+
+
+def apply_groupby(keys, values, ops, method='hash'):
+    """
+    Apply aggregations *ops* on *values*, grouping by *keys*.
+
+    Parameters
+    ----------
+    keys : list of Columns
+    values : list of Columns
+    ops : str or list of str
+        Aggregation to be performed for each column in *values*
+
+    Returns
+    -------
+    result : tuple of list of Columns
+        keys and values of the result
+    """
+    if len(values) == 0:
+        return (keys, [])
+
+    cdef pair[cudf_table, cudf_table] result
+    cdef cudf_table *c_keys_table = table_from_columns(keys)
+    cdef cudf_table *c_values_table = table_from_columns(values)
+    cdef vector[hash_groupby.operators] c_ops
+
+    num_values_cols = len(values)
+    for i in range(num_values_cols):
+        if isinstance(ops, str):
+            c_ops.push_back(agg_names[ops])
+        else:
+            c_ops.push_back(agg_names[ops[i]])
+
+    cdef hash_groupby.Options *options = new hash_groupby.Options()
+
+    result = hash_groupby.groupby(
+        c_keys_table[0],
+        c_values_table[0],
+        c_ops,
+        deref(options)
+    )
+
+    del options
+
+    result_key_cols = columns_from_table(&result.first)
+    result_value_cols = columns_from_table(&result.second)
+
+    for i, inp_key_col in enumerate(keys):
+        if isinstance(inp_key_col, CategoricalColumn):
+            result_key_cols[i] = CategoricalColumn(
+                data=result_key_cols[i].data,
+                mask=result_key_cols[i].mask,
+                categories=inp_key_col.cat().categories,
+                ordered=inp_key_col.cat().ordered
+            )
+
+    for i, inp_value_col in enumerate(values):
+         if isinstance(inp_value_col, CategoricalColumn):
+             result_value_cols[i] = CategoricalColumn(
+                 data=result_value_cols[i].data,
+                 mask=result_value_cols[i].mask,
+                 categories=inp_value_col.cat().categories,
+                 ordered=inp_value_col.cat().ordered
+             )
+
+    return (result_key_cols, result_value_cols)
+
+
+def agg(groupby_class, args):
+    """ Invoke aggregation functions on the groups.
+
+    Parameters
+    ----------
+    groupby_class : :class:`~cudf.groupby.Groupby`
+        Instance of :class:`~cudf.groupby.Groupby`.
+    args : dict, list, str, callable
+        - str
+            The aggregate function name.
+        - list
+            List of *str* of the aggregate function.
+        - dict
+            key-value pairs of source column name and list of
+            aggregate functions as *str*.
+
+    Returns
+    -------
+    result : DataFrame
+
+    Notes
+    -----
+    Since multi-indexes aren't supported aggregation results are returned
+    in columns using the naming scheme of `aggregation_columnname`.
```

Comment:
"Since multi-indexes aren't supported at the C++ level, aggregation results are returned..."

Reply:
Is this true anymore...? This seems reminiscent of before we made the decision to support MultiIndex and MultiColumn.

Reply:
In a sense even C++ supports MultiIndexes since it supports multiple keys in this aggregation method. To me all that MI means is how the dataframe is constructed and represented after groupby.

Reply:
@thomcom This section of the documentation can be removed, yes?

Reply:
@shwina Nuke it.